lib/network-layer/src/Cardano/Wallet/Network/Logging/Aggregation.hs

```diff
@@ -0,0 +1,207 @@
+      where
+        syncStatus = case progress of
+            Rearview NotResponding Ready ->
+                "In sync."
+            Rearview Ready Ready ->
+                "Still in sync."
+            Rearview NotResponding NotResponding ->
+                "Still not syncing."
+            Rearview (Syncing _p) Ready ->
+                "In sync!"
+            Rearview Ready (Syncing p) ->
+                "Fell out of sync (" <> (pretty p) <> ")"
+            Rearview _ (Syncing p) ->
+                "Syncing (" <> (pretty p) <> ")"
+            Rearview past_ NotResponding ->
+                "Not responding. Previously " <> (pretty past_) <> "."
+        stats =
+            mconcat
+                [ "Applied " <> pretty (using (-) b) <> " blocks, "
+                , pretty (using (-) r) <> " rollbacks "
+                , "in the last " <> pretty (using diffUTCTime t) <> ". "
+                , "Current tip is " <> pretty (current tip) <> "."
+                ]
+          where
+            using f x = f (current x) (past x)
+
+        sevExpl =
+            maybe
+                ""
+                (\x -> " (" <> x <> ")")
+                (snd $ explainedSeverityAnnotation st)
+
+-- NOTE: Here we check if the sync progress is going backwards, which
+-- would be a sign the wallet is overloaded (or rollbacks)
+--
+-- But this check might be in the wrong place. Might be better to
+-- produce new logs from inside the updateStats function and immeditely
+-- warn there.
+explainedSeverityAnnotation :: FollowStats Rearview -> (Severity, Maybe Text)
+explainedSeverityAnnotation s
+    | progressMovedBackwards = (Warning, Just "progress decreased")
+    | noBlocks && notRestored = (Warning, Just "not applying blocks")
+    | nowInSync = (Notice, Nothing)
+    | otherwise = (Info, Nothing)
+  where
+    progressMovedBackwards = current (prog s) < past (prog s)
+    nowInSync = current (prog s) == Ready && past (prog s) < Ready
+    notRestored = current (prog s) /= Ready
+    noBlocks = (current (blocksApplied s) - past (blocksApplied s)) <= 0
+
+instance HasSeverityAnnotation (FollowStats Rearview) where
+    getSeverityAnnotation = fst . explainedSeverityAnnotation
+
+-- | Update the 'TMVar' holding the 'FollowStats'@ @'Rearview'
+-- to forget the 'past' values and replace them with the 'current' ones.
+-- Also update the time and sync process.
+flushStats
+    :: UTCTime
+    -> (SlotNo -> IO SyncProgress)
+    -> StrictTMVar IO (FollowStats Rearview)
+    -> IO (FollowStats Rearview)
+flushStats t calcSyncProgress var = do
+    s <- atomically $ takeTMVar var
+    p <- calcSyncProgress $ pseudoSlotNo $ current $ localTip s
+    let s' =
+            s{time = overCurrent (const t) (time s)}
+                { prog = overCurrent (const p) (prog s)
+                }
+    atomically $ putTMVar var $ hoistStats forgetPast s'
+    return s'
+  where
+    forgetPast (Rearview _past curr) = initRearview curr
+
+-- See NOTE [PointSlotNo]
```

Comment:
is it here ? If not then you might expand on that

Reply:
I copied over. As always I have a bad feeling about this notes. But thanks I didn't spot it :+1: